src/engine/RenderSystem.cpp

```diff
@@ -0,0 +1,145 @@
+#include "engine/RenderSystem.hpp"
+
+#include "util/GLDebug.hpp"
+#include <array>
+
+namespace lei3d {
+
+void RenderSystem::initialize(int width, int height) {
+    scwidth = width;
+    scheight = height;
+
+    forwardShader = Shader("./data/shaders/forward.vert", "./data/shaders/forward.frag");
+    postprocessShader = Shader("./data/shaders/screenspace_quad.vert", "./data/shaders/postprocess.frag");
+
+    glGenVertexArrays(1, &dummyVAO);
+
+    glGenFramebuffers(1, &FBO);
+    glBindFramebuffer(GL_DRAW_FRAMEBUFFER, FBO);
+
+    glGenTextures(1, &rawTexture);
+    glGenTextures(1, &saturationMask);
+    glGenTextures(1, &depthStencilTexture);
+    glGenTextures(1, &finalTexture);
+
+    // lighting pass
+    glBindTexture(GL_TEXTURE_2D, rawTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_RGB, GL_FLOAT, nullptr);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, rawTexture, 0);
+
+    // saturation buffer
+    glBindTexture(GL_TEXTURE_2D, saturationMask);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_R32F, width, height, 0, GL_RED, GL_FLOAT, nullptr);   // single channel for one float values
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT1, GL_TEXTURE_2D, saturationMask, 0);
+
+    // depth map
+    glBindTexture(GL_TEXTURE_2D, depthStencilTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH32F_STENCIL8, width, height, 0, GL_DEPTH_STENCIL, GL_FLOAT_32_UNSIGNED_INT_24_8_REV, nullptr);   // probably won't need stencil
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_TEXTURE_2D, depthStencilTexture, 0);
+
+    // post process pass target
+    glBindTexture(GL_TEXTURE_2D, finalTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGB, GL_FLOAT, nullptr);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT2, GL_TEXTURE_2D, finalTexture, 0);
```

Comment:
Err, I understand that these steps are certainly procedural in nature, but I don't understand how these gl calls relate to each other procedurally. Very very cool and clever though.